src/storage/engine/btree_engine.rs

```diff
@@ -0,0 +1,644 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+use std::collections::Bound::{self, Excluded, Included, Unbounded};
+use std::fmt::{self, Debug, Formatter};
+use std::sync::{Arc, RwLock};
+
+use kvproto::kvrpcpb::Context;
+
+use raftstore::store::engine::IterOption;
+use storage::engine::{
+    Callback as EngineCallback, CbContext, Cursor, Engine, Error as EngineError, Iterator, Modify,
+    Result as EngineResult, ScanMode, Snapshot,
+};
+use storage::{CfName, Key, Value, CF_DEFAULT, CF_LOCK, CF_WRITE};
+
+#[derive(Clone)]
+pub struct BTreeEngine {
+    cf_names: Vec<CfName>,
+    cf_contents: Vec<Arc<RwLock<BTreeMap<Key, Value>>>>,
+}
+
+impl BTreeEngine {
+    pub fn new(cfs: &[CfName]) -> Self {
+        let mut cf_names = vec![];
+        let mut cf_contents = vec![];
+
+        // create default cf if missing
+        if cfs.iter().find(|&&c| c == CF_DEFAULT).is_none() {
+            cf_names.push(CF_DEFAULT);
+            cf_contents.push(Arc::new(RwLock::new(BTreeMap::new())))
+        }
+
+        for cf in cfs.iter() {
+            cf_names.push(*cf);
+            cf_contents.push(Arc::new(RwLock::new(BTreeMap::new())))
+        }
+
+        Self {
+            cf_names,
+            cf_contents,
+        }
+    }
+
+    pub fn default() -> Self {
+        let cfs = &[CF_WRITE, CF_DEFAULT, CF_LOCK];
+        Self::new(cfs)
+    }
+
+    pub fn get_cf(&self, cf: CfName) -> Arc<RwLock<BTreeMap<Key, Value>>> {
+        let index = self.cf_names.iter().position(|&c| c == cf);
+        match index {
+            None => unreachable!(
+                "Not exist CF:[{}]! Please create it by BTreeEngine::new(&[CfName])!",
+                cf
+            ),
+            Some(i) => self.cf_contents[i].clone(),
+        }
+    }
+}
+
+impl Engine for BTreeEngine {
+    type Iter = BTreeEngineIterator;
+    type Snap = BTreeEngineSnapshot;
+
+    fn async_write(
+        &self,
+        _ctx: &Context,
+        modifies: Vec<Modify>,
+        cb: EngineCallback<()>,
+    ) -> EngineResult<()> {
+        if modifies.is_empty() {
+            return Err(EngineError::EmptyRequest);
+        }
+        cb((CbContext::new(), write_modifies(&self, modifies)));
+
+        Ok(())
+    }
+    fn async_snapshot(&self, _ctx: &Context, cb: EngineCallback<Self::Snap>) -> EngineResult<()> {
+        cb((CbContext::new(), Ok(BTreeEngineSnapshot::new(&self))));
```

Comment:
Is this snapshot really snapshot? For example, when a snapshot is created, even if later there are new modifications the content of the snapshot won't change.

Reply:
No, it isn't. I implement the `async_snapshot()` because of the `Engine` trait, and the BTreeMapEngine only has few features in the early stage.

Reply:
If the snapshot is not really a snapshot, I guess it will affect the correctness of some tests 🤔

Reply:
Of cause it will, so be careful to use it.😂